core/src/apps/bitcoin/sign_tx/layout.py

```diff
@@ -84,6 +96,41 @@ async def confirm_decred_sstx_submission(
     )
 
 
+async def confirm_payment_request(
+    ctx: wire.Context,
+    msg: TxAckPaymentRequest,
+    coin: CoinInfo,
+    amount_unit: AmountUnit,
+) -> bool:
+    memo_texts = []
+    for memo in msg.memos:
+        if memo.type == MemoType.UTF8_TEXT:
+            memo_texts.append(memo.data.decode())
+        elif memo.type == MemoType.COIN_PURCHASE:
+            assert memo.amount is not None  # checked by sanitizer
+            assert memo.coin_name is not None  # checked by sanitizer
+            memo_coin = coininfo.by_name(memo.coin_name)
+            memo_texts.append(
+                "Buying "
+                + format_coin_amount(memo.amount, memo_coin, amount_unit)
+                + "."
+            )
+        else:
+            raise wire.DataError("Invalid memo type in payment request memo.")
+
+    if msg.amount is None:
```

Comment:
i'd prefer this to be validated in `BasicApprover.add_payment_request()`, and only asserted here

Reply:
The rationale behind this is that if someone adds a new memo type and forgets to implement it in the approver or more specifically in `PaymentRequestVerifier`, then the signature verification will fail, so there is no need to check it there. However, if they forget to implement it in the layouts, then the memo won't get shown to the user and the signing process will succeed. The assertion will only uncover it if the implementer also adds a device test. It would have to be an `ensure()` rather than `assert`.

Reply:
i specifically mean `msg.amount`, which the protobuf description claims is required for non-coinjoin

Reply:
Done as part of https://github.com/trezor/trezor-firmware/pull/2069/commits/94cef9693a8a69b1753eeae69e894f6bd52b9833